advanced/Exercise_2.md

```diff
@@ -0,0 +1,155 @@
+
+Now we get the correct answer: 3.14. So we know that somewhere between the first commit and the last commit, the script has stopped working. Git bisect will help us to figure out at exactly which commit this happened.
+
+Before we move on, let's switch our repository back to the final commit.  
+
+```plaintext
+git switch -
+```
+
+## Use git bisect manually to find the bad commit <a name="manual"></a>
+
+Initiate a git bisect session.
+
+```plaintext
+git bisect start
+```
+
+Next, you need to tell git bisect the commit you know where the code is good, and the commit you know where the code is bad. If you omit the commit SHA number, then git assumes you are referring to the current commit.  
+
+Tell git the current commit is bad.
+
+```plaintext
+git bisect bad
+```
+
+Tell git the first commit is good.
+
+```plaintext
+git bisect good f0ea950
+```
+
+Once you have run this command, git has enough information to start the bisection. It automatically finds the commit that is halfway in between the good commit and the bad commit, and checks it out for you. All you need to do is test the current commit, and then tell git whether the code is good or bad at this point.  
+
+```plaintext
+python get_pi.py
+```
+
+In this first case, you should see that the code gives 3.57, which is the incorrect answer. Let's inform git of this.
+
+```plaintext
+git bisect bad
+```
+
+Once you run this command, git will recalculate the halfway point between the new good and bad commits and check out a new commit for you to test. Continue by testing the script using `python get_pi.py` and then telling git whether the commit gives the correct result (`git bisect good`), or the incorrect result (`git bisect bad`). Keep repeating this process until you get a message from git identifying the exact commit where the code breaks. You can check the correct solution below.
+
+<details>
+  <summary>Click here to see the solution</summary>
+
+  The code breaks on commit number 137.
+</details>
+
+You should reset the repository to it's original state before you try any other git commands.
+
+```plaintext
+git bisect reset
+```
+
+## Use git bisect run to automatically find the bad commit <a name="automate"></a>
+
+We can also automate this process to make it easier. We just need a testing script that will return 0 if the code is correct, and an error value between 1 and 127 (excluding 125) if the code is incorrect. The error value 125 is reserved for when the code cannot be tested, and tells git bisect to skip the current commit. You could use whatever language you like to write such a script. If you are so inclined, you can try to write such a script yourself now. If not, you can use the python script provided below. If you do so, make sure you read through the python script so that you understand what it does.    
+
+<details>
+  <summary>Click here to see contents of the code checking script</summary>
+
+  ```plaintext
+  import subprocess
+  import numpy
+  import sys
+
```

Comment:
We should mention at the beginning that the participants need to set-up their python environment including those modules. Potentially add those to https://github.com/kosterried/git-course/blob/main/requirements.txt.